src/cmd_all/src/single_node.rs

```diff
@@ -65,41 +39,74 @@ pub struct SingleNodeOpts {
-    format!(
-        "sqlite://{}/meta_store/single_node.db?mode=rwc",
-        store_directory
-    )
+/// # Node-Specific Options
+///
+/// ## Which node-specific options should be here?
+///
+/// 1. An option doesn't make sense for single-node deployment should not be here.
+///    - e.g. `meta_addr` will be automatically set to localhost.
+/// 2. An option that is hidden from help message and can be set from config file
+///    - e.g. `etcd_endpoints` is encouraged to be set from config file or environment variables.
+/// 3. An option that is only for cloud deployment should not be here.
+///    - e.g. `proxy_rpc_endpoint` is used in cloud deployment.
+///
+/// ## How to add an new option here?
+///
+/// Options for the other nodes are defined with following convention:
+///
+/// 1. The option name is either the same as the node name or with a prefix to indicate the node type.
+/// 2. The option doesn't have a default value and must be `Option<T>`, so that the default value in the node's `Opts` struct can be used.
+/// 3. The option doesn't need to read from environment variables, which will be done in the node's `Opts` struct.
+#[derive(Eq, PartialOrd, PartialEq, Debug, Clone, Parser)]
+pub struct NodeSpecificOpts {
+    // ------- Compute Node Options -------
+    /// Total available memory for the compute node in bytes. Used by both computing and storage.
+    #[clap(long)]
+    pub total_memory_bytes: Option<usize>,
+
+    /// The parallelism that the compute node will register to the scheduler of the meta service.
+    #[clap(long)]
+    pub parallelism: Option<usize>,
+
+    // ------- Frontend Node Options -------
+    /// The address that this service listens to.
+    /// Usually the localhost + desired port.
+    #[clap(long)]
+    pub listen_addr: Option<String>,
+
+    // ------- Meta Node Options -------
+    /// The HTTP REST-API address of the Prometheus instance associated to this cluster.
+    /// This address is used to serve PromQL queries to Prometheus.
+    /// It is also used by Grafana Dashboard Service to fetch metrics and visualize them.
+    #[clap(long)]
+    pub prometheus_endpoint: Option<String>,
+
+    /// The additional selector used when querying Prometheus.
+    ///
+    /// The format is same as PromQL. Example: `instance="foo",namespace="bar"`
+    #[clap(long)]
+    pub prometheus_selector: Option<String>,
+
+    // ------- Compactor Node Options -------
+    #[clap(long)]
+    pub compaction_worker_threads_number: Option<usize>,
 }
 
-pub fn make_single_node_state_store_url(store_directory: &String) -> String {
-    format!("hummock+fs://{}/state_store", store_directory)
-}
+pub fn map_single_node_opts_to_standalone_opts(opts: SingleNodeOpts) -> ParsedStandaloneOpts {
+    // Parse from empty strings to get the default values.
+    // Note that environment variables will be used if they are set.
+    let empty_args = vec![] as Vec<String>;
```

Comment:
I suppose this means that all default values are for `single_node` mode.

Other deployment modes will override the values as needed.

Reply:
`advertise-addr` is trivial here, because it will be forced to override at L163.

https://github.com/risingwavelabs/risingwave/blob/f631e83b2ae8cff6889e04ada291248ebdf0dfc4/src/cmd_all/src/single_node.rs#L163

Actually, this is the only required argument we have (according to the [help messages](https://paste.ubuntu.com/p/jm2QFcVNPK/)), and I think it may not be intended. We may just give it a default value as other arguments.

Furthermore, assuming that we have a necessary required arguments someday, we should also let it be a required argument for single-node mode.